File: Algorithms/p12_Integer_to_Roman.py
class Solution(object):
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        """
        res = []
        
        d = (num/1000)%10
        if d>=1 and d<=3:
            res.append('M'*d)
        
        d = (num/100)%10
        if d>=1 and d<=3:
            res.append('C'*d)
        elif d==4:
            res.append('CD')
        elif d>=5 and d<9:
            res.append('D'+'C'*(d-5))
        elif d==9:
            res.append('CM')
        
        d = (num/10)%10
        if d>=1 and d<=3:
            res.append('X'*d)
        elif d==4:
            res.append('XL')
        elif d>=5 and d<9:
            res.append('L'+'X'*(d-5))
        elif d==9:
            res.append('XC')
        
        d = num%10
        if d>=1 and d<=3:
            res.append('I'*d)
        elif d==4:
            res.append('IV')
        elif d>=5 and d<9:
            res.append('V'+'I'*(d-5))
        elif d==9:
            res.append('IX')
        
        return ''.join(res)

File: Algorithms/p12_Integer_to_Roman.cpp
class Solution {
public:
    string intToRoman(int num) {
        string res;
        switch((num/1000)%10){
            case 1: res += "M";   break;
            case 2: res += "MM";  break;
            case 3: res += "MMM"; break;
        }
        switch((num/100)%10){
            case 1: res += "C";    break;
            case 2: res += "CC";   break;
            case 3: res += "CCC";  break;
            case 4: res += "CD";   break;
            case 5: res += "D";    break;
            case 6: res += "DC";   break;
            case 7: res += "DCC";  break;
            case 8: res += "DCCC"; break;
            case 9: res += "CM";   break;
        }
        switch((num/10)%10){
            case 1: res += "X";    break;
            case 2: res += "XX";   break;
            case 3: res += "XXX";  break;
            case 4: res += "XL";   break;
            case 5: res += "L";    break;
            case 6: res += "LX";   break;
            case 7: res += "LXX";  break;
            case 8: res += "LXXX"; break;
            case 9: res += "XC";   break;
        }
        switch(num%10){
            case 1: res += "I";    break;
            case 2: res += "II";   break;
            case 3: res += "III";  break;
            case 4: res += "IV";   break;
            case 5: res += "V";    break;
            case 6: res += "VI";   break;
            case 7: res += "VII";  break;
            case 8: res += "VIII"; break;
            case 9: res += "IX";   break;
        }
        return res;
    }
};
